src/game/npc/list.rs

```diff
@@ -191,23 +167,26 @@ impl<'a> NPCListMutableAliveIterator<'a> {
 }
 
 impl<'a> Iterator for NPCListMutableAliveIterator<'a> {
-    type Item = &'a mut NPC;
+    type Item = &'a RefCell<NPC>;
 
     fn next(&mut self) -> Option<Self::Item> {
         loop {
             if self.index >= self.map.max_npc.get() {
                 return None;
             }
 
-            let item = unsafe { self.map.npcs_mut().get_mut(self.index as usize) };
+            let item = self.map.npcs.get(self.index as usize);
             self.index += 1;
 
             match item {
                 None => {
                     return None;
                 }
-                Some(ref npc) if npc.cond.alive() => {
-                    return item;
+                // XXX: BEWARE, obscure logic bugs might appear if the user expects mutably-borrowed objects to be returned here!
```

Comment:
this is exactly the behavior we don't want here, even if it's unsound

Reply:
do you think that NPCList could be refactored to return `Rc<RefCell<NPC>>` or something similar?

Reply:
> do you think that NPCList could be refactored to return `Rc<RefCell<NPC>>` or something similar?

That won't fix the problem, which is that it is undefined behavior in Rust to create two mutable references to any value. The current engine creates a second mutable reference, such as to peek at the "alive" flag, and it seems to work, but in Rust this is "Your computer might explode" levels of forbidden behavior.

I've asked for feedback on the Rust subreddit here. https://www.reddit.com/r/rust/comments/1hnx9nm/improving_safety_in_doukutsurs_npc_list_request/